Mathlib/Tactic/Simps/Basic.lean

```diff
@@ -445,61 +446,104 @@ def projectionsInfo (l : List ProjectionData) (pref : String) (str : Name) : Mes
   let toPrint := MessageData.joinSep toPrint ("\n" : MessageData)
   m! "{pref} {str}:\n{toPrint}"
 
+open private mkBaseProjections from Lean.Elab.App
+
+/-- Elaborate the projection notation `$e.$projName` where `e` is an expression of a structure
+application `strName`. `e` might rely on universe variables `univs`.
+Note: `$e.$projName` is not applied to further (implicit) arguments. -/
+def elabProjectionNotation (univs : List Name) (e : Expr) (strName projName : Name) :
+  MetaM Expr :=
+TermElabM.run' (s := {levelNames := univs}) <| do
```

Comment:
```suggestion
TermElabM.run' (s := {levelNames := univs}) do
```
Oof, this isn't pretty.  I think it would be nicer to stay in `MetaM` entirely, and avoid accessing all these elaborator internals.

Building a non-subobject projection by hand shouldn't be too hard, I think we can just copy the params+univ params:
```lean
def mkDirectProjection (e : Expr) (fieldName : Name) : MetaM Expr := do
  let ty ← whnf (←inferType e)
  let .const structName us := ty.getAppFn | ...
  let some projName := getProjFnForField? (← getEnv) structName fieldName | ...
  -- also check out Meta.mkProjFn
  return mkAppN (.const projName us) (ty.getAppArgs.push e)
```

and then we could extend that to subobjects pretty easily:
```lean
def mkProjection (e : Expr) (fieldName : Name) : MetaM Expr := do
  let .const structName _ := (← whnf (←inferType e)).getAppFn | ..
  let some baseStruct := findField? (← getEnv) structName fieldName | ..
  let mut e := e
  for subobj in (getPathToBaseStructure? (← getEnv) baseStruct structName).get! do
    e ← mkDirectProjection e subobj
  mkDirectProjection e fieldName
```

Reply:
I was doing it manually before, and I got the remark that I was re-implementing parts of the elaborator, and it was maybe better to reuse it. But I'm happy to do it manually again.